### PR TITLE
`IndexShard.routingEntry` should only be updated once all internal state is ready

### PR DESCRIPTION
**Pull request by bleskes**
The routing entry is used by external components to check whether the shard is ready to perform as primary. Most notably, the peer recovery source handler delays recoveries until the shard routing entry says the shard is ready. 

When a shard is promoted to primary, we currently update the shard's routing entry before we finish all the work relating to the promotion. This can cause recoveries to fail later on because the `GlobalCheckpointTracker` isn't set (yet) to primary mode. 

This commit fixes this issue by updating the routing entry last.

### CODE REVIEW

**Review by s1monw — APPROVED** [review]
LGTM

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 1]
```diff
@@ -524,6 +523,8 @@ public void onFailure(Exception e) {
                     latch.countDown();
                 }
             }
+            // set this last, once we finished updating all internal state.
+            this.shardRouting = newRouting;
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I think that in the case of a primary this should be published before we countdown the latch that lets the operations that are run under the permit block proceed. At that point we will have published the new primary term but not the new routing entry, and this shard will be used for example in the resync operations. Am I missing something here?

**Review by jasontedor — CHANGES_REQUESTED** [review]
I left one question.

**Review by jasontedor — APPROVED** [review]
Assuming green CI, LGTM.

File: core/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 1 above)
Comment (bleskes) [reply core/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: good catch. Thanks.

**Comment by bleskes** [comment]
Thx @s1monw , @jasontedor

**Pull request merged**